flink-ml-servable-core/src/main/java/org/apache/flink/ml/servable/api/DataFrame.java

```diff
@@ -0,0 +1,132 @@
+ */
+@PublicEvolving
+public class DataFrame {
+
+    private final List<String> columnNames;
+    private final List<DataType> dataTypes;
+    private final List<Row> rows;
+
+    /**
+     * The given columnNames and dataTypes should be mutable in order for TransformerServable
+     * classes to update DataFrame with the serving results.
+     */
+    public DataFrame(List<String> columnNames, List<DataType> dataTypes, List<Row> rows) {
+        int numColumns = columnNames.size();
+        if (dataTypes.size() != numColumns) {
+            throw new IllegalArgumentException(
+                    String.format(
+                            "The number of data types %d is different from the number of column names %d.",
+                            dataTypes.size(), numColumns));
+        }
+        for (Row row : rows) {
+            if (row.size() != numColumns) {
+                throw new IllegalArgumentException(
+                        String.format(
+                                "The row size %d is different from the number of column names %d.",
+                                row.size(), numColumns));
+            }
+        }
+
+        this.columnNames = columnNames;
+        this.dataTypes = dataTypes;
+        this.rows = rows;
+    }
+
+    /** Returns a list of the names of all the columns in this DataFrame. */
+    public List<String> getColumnNames() {
+        return columnNames;
+    }
+
+    /**
+     * Returns the index of the column with the given name.
+     *
+     * @throws IllegalArgumentException if the column is not present in this table
+     */
+    public int getIndex(String name) {
+        int index = columnNames.indexOf(name);
+        if (index == -1) {
+            throw new IllegalArgumentException("Failed to find the column with the given name.");
+        }
+        return index;
+    }
+
+    /**
+     * Returns the data type of the column with the given name.
+     *
+     * @throws IllegalArgumentException if the column is not present in this table
+     */
+    public DataType getDataType(String name) {
+        int index = getIndex(name);
+        return dataTypes.get(index);
+    }
+
+    /**
+     * Adds to this DataFrame a column with the given name, data type, and values.
+     *
+     * @throws IllegalArgumentException if the number of values is different from the number of
+     *     rows.
+     */
+    public DataFrame addColumn(String columnName, DataType dataType, List<Object> values) {
+        if (values.size() != rows.size()) {
+            throw new RuntimeException(
+                    String.format(
+                            "The number of values %d is different from the number of rows %d.",
+                            values.size(), rows.size()));
+        }
+        columnNames.add(columnName);
+        dataTypes.add(dataType);
+
+        int rowSize = -1;
+        Iterator<Object> iter = values.iterator();
+        for (Row row : rows) {
+            if (rowSize < 0) {
```

Comment:
The check for row size here seems uncessary since we already checked them in the constructor.

Reply:
Thanks for catching this! It is fixed now.